src/Type/Php/ArrayReplaceFunctionDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,67 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Reflection\ParametersAcceptorSelector;
+use PHPStan\Type\Accessory\NonEmptyArrayType;
+use PHPStan\Type\ArrayType;
+use PHPStan\Type\GeneralizePrecision;
+use PHPStan\Type\Type;
+use PHPStan\Type\TypeCombinator;
+use PHPStan\Type\TypeUtils;
+use PHPStan\Type\UnionType;
+
+class ArrayReplaceFunctionDynamicReturnTypeExtension implements \PHPStan\Type\DynamicFunctionReturnTypeExtension
+{
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return $functionReflection->getName() === 'array_replace';
+	}
+
+	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): Type
+	{
+		if (!isset($functionCall->getArgs()[0])) {
+			return ParametersAcceptorSelector::selectSingle($functionReflection->getVariants())->getReturnType();
+		}
+
+		$keyTypes = [];
+		$valueTypes = [];
+		$nonEmpty = false;
+		foreach ($functionCall->getArgs() as $arg) {
+			$argType = $scope->getType($arg->value);
+			if ($arg->unpack) {
+				$argType = $argType->getIterableValueType();
+				if ($argType instanceof UnionType) {
+					foreach ($argType->getTypes() as $innerType) {
+						$argType = $innerType;
```

Comment:
I don't understand this part - the last inner type in the union will win in the `$argType` variable?

Reply:
I know the same code is in ArrayMergeFunctionDynamicReturnTypeExtension - it also puzzles me there.

Reply:
:)

Reply:
Let's try and see what happens :)

Reply:
I re-added the fixes and added a limit for the array. I also added the fix for "array_merge" here, so we can see the differences between the implementations in one diff.